<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>model Module &mdash; Pyvolve  documentation</title>
    
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="shortcut icon" href="_static/pyvolve_favicon.ico"/>
    <link rel="top" title="Pyvolve  documentation" href="index.html" />
    <link rel="up" title="Python modules in pyvolve" href="modules.html" />
    <link rel="next" title="partition Module" href="partition.html" />
    <link rel="prev" title="newick Module" href="newick.html" /> 
  </head>
  <body role="document">
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">Pyvolve  documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="newick.html" title="newick Module"
             accesskey="P">previous</a> |
          <a href="partition.html" title="partition Module"
             accesskey="N">next</a> |
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-model">
<span id="model-module"></span><h1><code class="docutils literal"><span class="pre">model</span></code> Module<a class="headerlink" href="#module-model" title="Permalink to this headline">¶</a></h1>
<p>Define evolutionary model objects.</p>
<dl class="class">
<dt id="model.Model">
<em class="property">class </em><code class="descclassname">model.</code><code class="descname">Model</code><span class="sig-paren">(</span><em>model_type</em>, <em>parameters=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#model.Model" title="Permalink to this definition">¶</a></dt>
<dd><p>This class defines evolutionary model objects.
All evolutionary models contain information about the substitution process (rate matrix) and information about rate heterogeneity.
Note that, in cases of rate heterogeneity, non-dN/dS models use a single rate matrix and model heterogeneity using discrete scaling factors and associated probabilities.
Alternatively, rate heterogeneity in dN/dS models is implemented using a set of matrices with distinct dN/dS values, and each matrix has an associated probability.</p>
<p>The Model class will construct an evolutionary model object which will be used to evolve sequence data.            
Instantiation requires a single positional argument (but a second one is recommended, read on!):</p>
<blockquote>
<div><ol class="arabic">
<li><p class="first"><strong>model_type</strong> is type of model (matrix) that is being used. These matrices are described explicitly in the matrix_builder module. Options include the following:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="83%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">model_type</th>
<th class="head">Notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>nucleotide</td>
<td>Arbitrary GTR</td>
</tr>
<tr class="row-odd"><td>JTT</td>
<td>Jones, Taylor, and Thornton 1994 (amino acids)</td>
</tr>
<tr class="row-even"><td>WAG</td>
<td>Whelan and Goldman 2002      (amino acids)</td>
</tr>
<tr class="row-odd"><td>LG</td>
<td>Le and Gascuel 2008        (amino acids)</td>
</tr>
<tr class="row-even"><td>MTMAM</td>
<td>Yang, Nielsen, and Hasagawa 1998   (amino acids)</td>
</tr>
<tr class="row-odd"><td>MTREV24</td>
<td>Adachi and Hasegawa  1996 (amino acids)</td>
</tr>
<tr class="row-even"><td>DAYHOFF</td>
<td>Dayhoff, Schwartz, and Orcutt  1978 (amino acids)</td>
</tr>
<tr class="row-odd"><td>AB</td>
<td>Mirsky, Kazandjian, and Anisimova 2015    (amino acids)</td>
</tr>
<tr class="row-even"><td>GY</td>
<td>Goldman and Yang 1994 (modified), Nielsen and Yang 1998</td>
</tr>
<tr class="row-odd"><td>MG</td>
<td>Muse and Gaut 1994</td>
</tr>
<tr class="row-even"><td>codon</td>
<td>Defaults to GY-style model</td>
</tr>
<tr class="row-odd"><td>ECM</td>
<td>Kosiol et al. 2007</td>
</tr>
<tr class="row-even"><td>mutsel</td>
<td>Halpern and Bruno 2008 (may also be used for nucleotides)</td>
</tr>
</tbody>
</table>
</div></blockquote>
</li>
</ol>
</div></blockquote>
<p>To use your own rate matrix (which you must create on your own), enter &#8220;custom&#8221; for the model_type argument, and provide the custom matrix (numpy array or list of lists) in the <strong>parameters</strong> dictionary with the key &#8220;matrix&#8221;. Please note that pyvolve stores nucleotides, amino acids, and codons in alphabetical order of their abbreviations:
*  Nucleotides: A, C, G, T
*  Amino acids: A, C, D, E, F, G, H, I, K, L, M, N, P, Q, R, S, T, V, W, Y
*  Codons:      AAA, AAC, AAG, AAT, ACA, ... TTG, TTT [note that stop codons should be <em>excluded</em>]</p>
<p>If you wish to evolve <em>custom states</em> (neither nucleotide, amino acids, nor codons), for instance to evolve characters, also include the key &#8220;code&#8221; in the parameters dictionary. The associated value should be a list of strings, e.g. [&#8220;0&#8221;, &#8220;1&#8221;, &#8220;2&#8221;], and the length of this list should be the same as a dimension of the square custom matrix provided. Note that this argument is not required if wish to evolve nucleotides, amino-acids, and/or codons.</p>
<p>Please be careful here - Pyvolve takes your matrix (mostly) at face-value (provided it has proper dimensions and rows sum to 0). In particular, the matrix will not be scaled!!!</p>
<p>A second positional argument, <strong>parameters</strong> may additionally be specified. This argument should be a dictionary of parameters pertaining to substitution process. Each individual evolutionary model will have its own parameters. Note that if this argument is not provided, default parameters for your selected model will be assigned. Note that this argument is <strong>required</strong> for mechanistic codon (dN/dS) models, as this rate ratio must be assigned!</p>
<dl class="docutils">
<dt>Optional keyword arguments include, </dt>
<dd><ol class="first last arabic simple">
<li><strong>scale_matrix</strong> = &lt;&#8217;yang&#8217;, &#8216;neutral&#8217;&gt;. This argument determines how rate matrices should be scaled. By default, all matrices are scaled according to Ziheng Yang&#8217;s approach, in which the mean substitution rate is equal to 1. However, for codon models (GY-style and MG-style), this scaling approach effectively causes sites under purifying selection to evolve at the same rate as sites under positive selection, which may not be desired. Thus, the &#8216;neutral&#8217; scaling option will allow for codon matrices to be scaled such that the mean rate of <em>neutral</em> subsitution is 1.</li>
<li><strong>name</strong>, the name for a Model object. Names are not needed in cases of branch homogeneity, but when there is <strong>branch heterogeneity</strong>, names are required to map the model to the model flags provided in the phylogeny.</li>
<li><strong>rate_factors</strong>, for specifying rate heterogeneity in nucleotide or amino acid models. This argument should be a list/numpy array of scalar factors for rate heterogeneity. Default: rate homogeneity.</li>
<li><strong>rate_probs</strong>, for specifying rate heterogeneity probabilities in nucleotide, amino acid, or codon models. This argument should be a list/numpy array of probabilities (which sum to 1!) for each rate category. Default: equal.</li>
<li><strong>alpha</strong>, for specifying rate heterogeneity in nucleotide or amino acid models if gamma-distributed heterogeneity is desired. The alpha shape parameter which should be used to draw rates from a discrete gamma distribution.</li>
<li><strong>num_categories</strong>, for specifying the number of gamma categories to draw for rate heterogeneity in nucleotide or amino acid models. Should be used in conjunction with the &#8220;alpha&#8221; parameter. Default: 4.</li>
<li><strong>pinv</strong>, for specifying a proportion of invariant sites when gamma heterogeneity is used. When specifying custom rate heterogeneity, a proportion of invariant sites can be specified simply with a rate factor of 0.</li>
<li><strong>save_custom_frequencies</strong>, for specifying a file name in which to save the state frequencies from a custom matrix. Pyvolve automatically computes the proper frequencies and will save them to a file named &#8220;custom_matrix_frequencies.txt&#8221;, and you can use this argument to change the file name. Note that this argument is really only relevant for custom models.</li>
</ol>
</dd>
</dl>
<dl class="method">
<dt id="model.Model.num_classes">
<code class="descname">num_classes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#model.Model.num_classes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the number of rate classes associated with a given model.</p>
</dd></dl>

<dl class="method">
<dt id="model.Model.assign_name">
<code class="descname">assign_name</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="headerlink" href="#model.Model.assign_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Assign name to a Model instance. 
In cases of branch/temporal homogeneity, names are unneeded.
However, in cases of <strong>branch heterogeneity, each model must be named</strong>. Names are used to map to model flags given in the phylogeny.
NOTE that name can also be assigned as a keyword argument when initializing a Model object.</p>
</dd></dl>

<dl class="method">
<dt id="model.Model.is_codon_model">
<code class="descname">is_codon_model</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#model.Model.is_codon_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if the model is a heterogeneous codon model and return False otherwise.</p>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <p class="logo"><a href="index.html">
            <img class="logo" src="_static/pyvolve_logo.png" alt="Logo"/>
          </a></p>
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Python modules in <cite>pyvolve</cite></a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="genetics.html"><code class="docutils literal"><span class="pre">genetics</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="newick.html"><code class="docutils literal"><span class="pre">newick</span></code> Module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href=""><code class="docutils literal"><span class="pre">model</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="partition.html"><code class="docutils literal"><span class="pre">partition</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="state_freqs.html"><code class="docutils literal"><span class="pre">state_freqs</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matrix_builder.html"><code class="docutils literal"><span class="pre">matrix_builder</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="evolver.html"><code class="docutils literal"><span class="pre">evolver</span></code> Module</a></li>
</ul>
</li>
</ul>

          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
            </form>
            <p class="searchtip" style="font-size: 90%">
                Enter search terms or a module, class or function name.
            </p>
          </div>

        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="newick.html" title="newick Module"
              >previous</a> |
            <a href="partition.html" title="partition Module"
              >next</a> |
            <a href="py-modindex.html" title="Python Module Index"
              >modules</a> |
            <a href="genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="_sources/model.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Stephanie J. Spielman.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>