

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>state_freqs Module &mdash; Pyvolve 1.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Pyvolve 1.0 documentation" href="index.html"/>
        <link rel="up" title="Python modules in pyvolve" href="modules.html"/>
        <link rel="next" title="matrix_builder Module" href="matrix_builder.html"/>
        <link rel="prev" title="partition Module" href="partition.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        

        
          <a href="index.html" class="icon icon-home"> Pyvolve
        

        
        </a>

        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Python modules in <cite>pyvolve</cite></a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="genetics.html"><code class="docutils literal"><span class="pre">genetics</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="newick.html"><code class="docutils literal"><span class="pre">newick</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="model.html"><code class="docutils literal"><span class="pre">model</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="partition.html"><code class="docutils literal"><span class="pre">partition</span></code> Module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href=""><code class="docutils literal"><span class="pre">state_freqs</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matrix_builder.html"><code class="docutils literal"><span class="pre">matrix_builder</span></code> Module</a></li>
<li class="toctree-l2"><a class="reference internal" href="evolver.html"><code class="docutils literal"><span class="pre">evolver</span></code> Module</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Pyvolve</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
          <li><a href="modules.html">Python modules in <cite>pyvolve</cite></a> &raquo;</li>
      
    <li><code class="docutils literal"><span class="pre">state_freqs</span></code> Module</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/state_freqs.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="module-state_freqs">
<span id="state-freqs-module"></span><h1><code class="docutils literal"><span class="pre">state_freqs</span></code> Module<a class="headerlink" href="#module-state_freqs" title="Permalink to this headline">¶</a></h1>
<p>This module will compute a vector of stationary frequencies.</p>
<dl class="class">
<dt id="state_freqs.StateFrequencies">
<em class="property">class </em><code class="descclassname">state_freqs.</code><code class="descname">StateFrequencies</code><span class="sig-paren">(</span><em>by</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#state_freqs.StateFrequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<blockquote>
<div>Parent class for stationary (state, equilibrium, etc.) frequency calculations.</div></blockquote>
<dl class="docutils">
<dt>Child classes include the following:</dt>
<dd><ol class="first last arabic">
<li><dl class="first docutils">
<dt><strong>EqualFrequencies</strong> (default)</dt>
<dd><ul class="first last simple">
<li>Sets frequencies as equal (i.e. 1/4 for all nucleotides if by=&#8217;nucleotide&#8217;, and so on.)</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>RandomFrequencies</strong> </dt>
<dd><ul class="first last simple">
<li>Computes (semi-)random frequency values for a given alphabet.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>CustomFrequencies</strong></dt>
<dd><ul class="first last simple">
<li>Computes frequencies based on a user-provided dictionary of frequencies.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>ReadFrequencies</strong> </dt>
<dd><ul class="first last simple">
<li>Computes frequencies from a sequence file. Contains an option to select specific columns from sequence file only, but this requires that the file is an alignemnt.</li>
</ul>
</dd>
</dl>
</li>
</ol>
</dd>
</dl>
<p>A single positional argument is required for all child classes.
This argument can take on three values: &#8220;nucleotide&#8221;, &#8220;amino_acid&#8221;, or &#8220;codon,&#8221; and it indicates <em>how</em> frequencies should be computed. These frequencies need not be the ultimate frequencies you want to compute. For example, it is possible to compute stationary frequencies in amino-acid space (via this argument) but ultimately return codon frequencies (using argument &#8220;type&#8221; in the .compute_frequencies() method, described below).</p>
<dl class="method">
<dt id="state_freqs.StateFrequencies.compute_frequencies">
<code class="descname">compute_frequencies</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#state_freqs.StateFrequencies.compute_frequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate and return a vector of state frequencies. At this stage, the StateFrequencies object must already have been initialized with the keyword argument by = &lt;amino_acid/codon/nucleotide&gt;.</p>
<p>Optional keyword arguments include,</p>
<blockquote>
<div><ol class="arabic simple">
<li><strong>type</strong> ( = &#8220;nucleotide&#8221;, &#8220;amino_acid&#8221;, or &#8220;codon&#8221;) represents the type of final frequencies to return. If not specified, the alphabet of returned frequencies will be that specified with the <strong>by</strong> keyword.</li>
<li><strong>savefile</strong> is a file name to which final frequencies may be saved. Output frequencies will be ordered alphabetically, i.e. A, C, G, T for nucleotides; A, C, D, E, etc.for amino acids; and AAA, AAC, AAG, AAT, ACA, etc. for codons.</li>
</ol>
</div></blockquote>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="state_freqs.EqualFrequencies">
<em class="property">class </em><code class="descclassname">state_freqs.</code><code class="descname">EqualFrequencies</code><span class="sig-paren">(</span><em>by</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#state_freqs.EqualFrequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#state_freqs.StateFrequencies" title="state_freqs.StateFrequencies"><code class="xref py py-class docutils literal"><span class="pre">state_freqs.StateFrequencies</span></code></a></p>
<p>This class may be used to compute equal state frequencies (amino = 1/20, codon = 1/61, nucleotide = 1/4).</p>
<blockquote>
<div><p>Required arguments include,</p>
<blockquote>
<div><ol class="arabic simple">
<li><strong>by</strong>. See parent class StateFrequencies for details.</li>
</ol>
</div></blockquote>
<p>Optional arguments include,</p>
<ol class="arabic simple">
<li><strong>restrict</strong>, a list (in which each element is a string) specifying which states should have non-zero frequencies. Default: all.</li>
</ol>
</div></blockquote>
<dl class="docutils">
<dt>Examples:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># Return 1/20 amino acid frequencies in the variable `frequencies`</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">EqualFrequencies</span><span class="p">(</span><span class="s">&quot;amino_acid&quot;</span><span class="p">)()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">contruct_frequencies</span><span class="p">()</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># Compute equal codon frequencies and convert to amino-acid space. `frequencies` will contain amino-acid frequencies.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">EqualFrequencies</span><span class="p">(</span><span class="s">&quot;codon&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">(</span><span class="nb">type</span> <span class="o">=</span> <span class="s">&quot;amino_acid&quot;</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># Compute equal amino acid frequencies, but allowing only certain amino acids to have non-zero frequencies</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">EqualFrequencies</span><span class="p">(</span><span class="s">&quot;amino_acid&quot;</span><span class="p">,</span> <span class="n">restrict</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;A&quot;</span><span class="p">,</span> <span class="s">&quot;G&quot;</span><span class="p">,</span> <span class="s">&quot;P&quot;</span><span class="p">,</span> <span class="s">&quot;T&quot;</span><span class="p">,</span> <span class="s">&quot;W&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="state_freqs.RandomFrequencies">
<em class="property">class </em><code class="descclassname">state_freqs.</code><code class="descname">RandomFrequencies</code><span class="sig-paren">(</span><em>by</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#state_freqs.RandomFrequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#state_freqs.StateFrequencies" title="state_freqs.StateFrequencies"><code class="xref py py-class docutils literal"><span class="pre">state_freqs.StateFrequencies</span></code></a></p>
<p>This class may be used to compute &#8220;semi-random&#8221; state frequencies. The resulting frequency distributions are not truly random, but are instead virtually flat distributions with some noise.</p>
<blockquote>
<div><p>Required arguments include,</p>
<blockquote>
<div><ol class="arabic simple">
<li><strong>by</strong>. See parent class StateFrequencies for details.</li>
</ol>
</div></blockquote>
<p>Optional arguments include,</p>
<ol class="arabic simple">
<li><strong>restrict</strong>, a list (in which each element is a string) specifying which states should have non-zero frequencies. Default: all.</li>
</ol>
</div></blockquote>
<dl class="docutils">
<dt>Examples:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># Return random amino acid frequencies in `frequencies` variable</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">RandomFrequencies</span><span class="p">(</span><span class="s">&quot;amino_acid&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>


<span class="gp">&gt;&gt;&gt; </span><span class="c"># Compute random amino acid frequencies, but allowing only certain amino acids to have non-zero frequencies</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">RandomFrequencies</span><span class="p">(</span><span class="s">&quot;amino_acid&quot;</span><span class="p">,</span> <span class="n">restrict</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;A&quot;</span><span class="p">,</span> <span class="s">&quot;G&quot;</span><span class="p">,</span> <span class="s">&quot;P&quot;</span><span class="p">,</span> <span class="s">&quot;T&quot;</span><span class="p">,</span> <span class="s">&quot;W&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="state_freqs.CustomFrequencies">
<em class="property">class </em><code class="descclassname">state_freqs.</code><code class="descname">CustomFrequencies</code><span class="sig-paren">(</span><em>by</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#state_freqs.CustomFrequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#state_freqs.StateFrequencies" title="state_freqs.StateFrequencies"><code class="xref py py-class docutils literal"><span class="pre">state_freqs.StateFrequencies</span></code></a></p>
<blockquote>
<div>This class may be used to compute frequencies directly from a user-provided python dictionary of frequencies.</div></blockquote>
<p>Required keyword arguments include,</p>
<blockquote>
<div><blockquote>
<div><ol class="arabic simple">
<li><strong>by</strong>. See parent class StateFrequencies for details.</li>
<li><strong>freq_dict</strong>, a dictionary of frequencies, in which keys are states (e.g. a codon key would be &#8216;ATC&#8217;, an amino acid key would be &#8216;W&#8217;, and a nucleotide key would be &#8216;T&#8217;), and values are float frequencies which sum to 1. Note that the keys in this dictionary must correspond to the <strong>by</strong> keyword provided. Any states not included in this dictionary are assumed to have an equal frequency. Hence, the dictionary values <em>MUST</em> sum to 1, and all states not included in this dictionary will be given a 0 frequency.</li>
</ol>
</div></blockquote>
<dl class="docutils">
<dt>Examples:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># custom random amino acid frequencies</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">CustomFrequencies</span><span class="p">(</span><span class="s">&quot;amino_acid&quot;</span><span class="p">,</span> <span class="n">freq_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;D&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;E&#39;</span><span class="p">:</span><span class="mf">0.3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># use amino-acid information to get custom codon frequencies (note: synonymous codons are assigned equal frequencies!)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">CustomFrequencies</span><span class="p">(</span><span class="s">&quot;amino_acid&quot;</span><span class="p">,</span> <span class="n">freq_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;F&#39;</span><span class="p">:</span><span class="mf">0.5</span><span class="p">,</span> <span class="s">&#39;W&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;D&#39;</span><span class="p">:</span><span class="mf">0.2</span><span class="p">,</span> <span class="s">&#39;E&#39;</span><span class="p">:</span><span class="mf">0.3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">(</span><span class="nb">type</span> <span class="o">=</span> <span class="s">&quot;codon&quot;</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># custom nucleotide frequencies with lots of GC bias</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">CustomFrequencies</span><span class="p">(</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">freq_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="mf">0.1</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="mf">0.45</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="mf">0.05</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span> <span class="mf">0.4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>
</pre></div>
</div>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="class">
<dt id="state_freqs.ReadFrequencies">
<em class="property">class </em><code class="descclassname">state_freqs.</code><code class="descname">ReadFrequencies</code><span class="sig-paren">(</span><em>by</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#state_freqs.ReadFrequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#state_freqs.StateFrequencies" title="state_freqs.StateFrequencies"><code class="xref py py-class docutils literal"><span class="pre">state_freqs.StateFrequencies</span></code></a></p>
<blockquote>
<div><p>This class may be used to compute frequencies directly from a specified sequence file. Frequencies may be computed globally (using entire file), or based on specific columns (i.e. site-specific frequencies), provided the file contains a sequence alignment.</p>
<dl class="docutils">
<dt>Required positional include, </dt>
<dd><ol class="first last arabic simple">
<li><strong>by</strong>. See parent class StateFrequencies for details.</li>
</ol>
</dd>
<dt>Required keyword arguments include, </dt>
<dd><ol class="first last arabic simple">
<li><strong>file</strong> is the file containing sequences from which frequencies will be computed. By default, this file is assumed to be in FASTA format, although you can specify a different format with the optional argument <strong>format</strong></li>
</ol>
</dd>
<dt>Optional keyword arguments include, </dt>
<dd><ol class="first last arabic simple">
<li><strong>format</strong> is the sequence file format (case-insensitive). Sequence files are parsed using Biopython, so any format they accept is accepted here (e.g. fasta, phylip, phylip-relaxed, nexus, clustal...)</li>
<li><strong>columns</strong> is a list of integers giving the column(s) which should be considered in frequency calculations. This list should be indexed <em>from 1</em>. If this argument is not provided, all positions in sequence file will be considered. Note that this argument is only possible for alignments!</li>
</ol>
</dd>
</dl>
</div></blockquote>
<dl class="docutils">
<dt>Examples:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># Compute amino acid frequencies globally from a sequence file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">ReadFrequencies</span><span class="p">(</span><span class="s">&quot;amino_acid&quot;</span><span class="p">,</span> <span class="nb">file</span> <span class="o">=</span> <span class="s">&quot;my_sequence_file.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># Compute amino acid frequencies globally from a sequence file, and then convert to codon frequencies (note: synonymous codons are assigned the same fitness!)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">ReadFrequencies</span><span class="p">(</span><span class="s">&quot;amino_acid&quot;</span><span class="p">,</span> <span class="nb">file</span> <span class="o">=</span> <span class="s">&quot;my_sequence_file.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">(</span><span class="nb">type</span> <span class="o">=</span> <span class="s">&quot;codon&quot;</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># Compute nucleotide frequencies from a specific range of columns (1-10, inclusive) from a nucleotide alignment file </span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">ReadFrequencies</span><span class="p">(</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">file</span> <span class="o">=</span> <span class="s">&quot;my_nucleotide_alignment.phy&quot;</span><span class="p">,</span> <span class="n">format</span> <span class="o">=</span> <span class="s">&quot;phylip&quot;</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">11</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="state_freqs.EmpiricalModelFrequencies">
<em class="property">class </em><code class="descclassname">state_freqs.</code><code class="descname">EmpiricalModelFrequencies</code><span class="sig-paren">(</span><em>model</em><span class="sig-paren">)</span><a class="headerlink" href="#state_freqs.EmpiricalModelFrequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>This class assigns state frequencies from a specified amino acid or codon empirical model (e.g. JTT, WAG, ECM...). The default frequencies (i.e. those given in each model&#8217;s original paper) for empirical models.</p>
<dl class="docutils">
<dt>The currently supported models include, </dt>
<dd><ol class="first last arabic simple">
<li><em>Amino acid</em>: JTT, WAG, LG</li>
<li><em>Codon</em>:      ECM(un)rest</li>
</ol>
</dd>
<dt>Required positional arguments include, </dt>
<dd><ol class="first last arabic simple">
<li><strong>model</strong> is empirical model of choice (case-insensitive). This argument should be specified as any of the following: JTT, WAG, LG, ECMrest, ECMunrest.</li>
</ol>
</dd>
<dt>Examples:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># Assign WAG frequencies</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">EmpiricalModelFrequencies</span><span class="p">(</span><span class="s">&quot;WAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># Assign ECMrest frequencies (ECM &quot;restricted&quot; model, in which only single nucleotide changes occur instantaneously)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_freqs</span> <span class="o">=</span> <span class="n">EmpiricalModelFrequencies</span><span class="p">(</span><span class="s">&quot;ecmrest&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">compute_frequencies</span><span class="p">()</span>
</pre></div>
</div>
</dd>
</dl>
<dl class="method">
<dt id="state_freqs.EmpiricalModelFrequencies.compute_frequencies">
<code class="descname">compute_frequencies</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#state_freqs.EmpiricalModelFrequencies.compute_frequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to return state frequencies. No arguments are needed.</p>
</dd></dl>

</dd></dl>

</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="matrix_builder.html" class="btn btn-neutral float-right" title="matrix_builder Module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="partition.html" class="btn btn-neutral" title="partition Module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, Stephanie J. Spielman.
    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>